### PR TITLE
mysql/decimal: remove read trait from decode

### PR DESCRIPTION
**Pull request by AndreMouche**
Hi,
This PR updates the decode function for`Decimal`. It accesses slice directly instead of using read trait since read trait is not sufficient enough.
@BusyJay @hicqu @breeswish PTAL

### CODE REVIEW

**Comment by BusyJay** [comment]
Any benchmark result?

**Comment by AndreMouche** [comment]
bench test 
```
running 2 tests
test decimal::bench_decode_decimal          ... bench:         250 ns/iter (+/- 35)
test decimal::bench_decode_decimal_read     ... bench:         291 ns/iter (+/- 165)
```

**Comment by AndreMouche** [comment]
benchmark result
```
test decimal::bench_decode_decimal              ... bench:         301 ns/iter (+/- 246)
test decimal::bench_decode_decimal_read         ... bench:         348 ns/iter (+/- 337)
```
test source code

```
#[bench]
fn bench_decode_decimal_read(b: &mut Bencher) {
    let cases = vec![
        ("-10.55", 4, 2),
        (
            "0.0123456789012345678912345",
            30,
            25,
            
        ),
        (".12345000098765", 30, 20),
        
        ("1234500009876.5", 30, 5), 
    ];
    let expect = vec!["-10.55","0.0123456789012345678912345","0.12345000098765000000","1234500009876.50000"];
    let mut buf = Vec::new();
    let mut res:Vec<Decimal> = Vec::with_capacity(cases.len());
    for (dec_str, prec, frac) in cases {
        let dec = dec_str.parse::<Decimal>().unwrap();
        buf.encode_decimal(&dec, prec, frac).unwrap();
        res.push(dec);
    }
    
    b.iter(|| {
       let mut slice = buf.as_slice();
       res[0] = slice.decode_decimal().unwrap();
       res[1] = slice.decode_decimal().unwrap();
       res[2] = slice.decode_decimal().unwrap();
       res[3] = slice.decode_decimal().unwrap();
    });

    for id in 0..res.len() {
        assert_eq!(res[id].to_string(),expect[id].to_owned());
    }
}

#[bench]
fn bench_decode_decimal(b: &mut Bencher) {
    let cases = vec![
        ("-10.55", 4, 2),
        (
            "0.0123456789012345678912345",
            30,
            25,
            
        ),
        (".12345000098765", 30, 20),
        
        ("1234500009876.5", 30, 5), 
    ];
    let expect = vec!["-10.55","0.0123456789012345678912345","0.12345000098765000000","1234500009876.50000"];
    let mut buf = Vec::new();
    let mut res:Vec<Decimal> = Vec::with_capacity(cases.len());
    for (dec_str, prec, frac) in cases {
        let dec = dec_str.parse::<Decimal>().unwrap();
        buf.encode_decimal(&dec, prec, frac).unwrap();
        res.push(dec);
    }
    
    b.iter(|| {
       let mut slice = buf.as_slice();
       res[0] = Decimal::decode(&mut slice).unwrap();
       res[1] = Decimal::decode(&mut slice).unwrap();
       res[2] = Decimal::decode(&mut slice).unwrap();
       res[3] = Decimal::decode(&mut slice).unwrap();
    });

    for id in 0..res.len() {
        assert_eq!(res[id].to_string(),expect[id].to_owned());
    }
}
```

File: src/coprocessor/codec/mysql/decimal.rs [hunk 1]
```diff
@@ -1746,30 +1744,36 @@ macro_rules! write_word {
 }
 
 macro_rules! read_word {
```
Comment (hicqu) [inline src/coprocessor/codec/mysql/decimal.rs]: Could we write it as a function?

**Comment by hicqu** [comment]
rest LGTM.

File: src/coprocessor/codec/mysql/decimal.rs (hunk 1 above)
Comment (AndreMouche) [reply src/coprocessor/codec/mysql/decimal.rs]: no, in this case, `macro_rules` has a better performance than function.

**Comment by AndreMouche** [comment]
friendly ping @BusyJay @breeswish

**Comment by breeswish** [comment]
LGTM

**Pull request merged**